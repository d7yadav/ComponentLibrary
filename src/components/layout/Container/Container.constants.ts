import type { ContainerMaxWidth, ContainerVariant } from './Container.types';

export const CONTAINER_VARIANTS: Record<ContainerVariant, ContainerVariant> = {
  fluid: 'fluid',
  fixed: 'fixed',
  constrained: 'constrained',
} as const;

export const CONTAINER_MAX_WIDTHS: Record<string, ContainerMaxWidth> = {
  xs: 'xs',
  sm: 'sm', 
  md: 'md',
  lg: 'lg',
  xl: 'xl',
  false: false,
} as const;

export const CONTAINER_BREAKPOINTS = {
  xs: 0,
  sm: 600,
  md: 900,
  lg: 1200,
  xl: 1536,
} as const;

export const CONTAINER_MAX_WIDTH_VALUES = {
  xs: 444,
  sm: 600,
  md: 900,
  lg: 1200,
  xl: 1536,
} as const;

export const CONTAINER_GUTTERS = {
  xs: 16,
  sm: 24,
  md: 32,
  lg: 32,
  xl: 32,
} as const;

export const CONTAINER_PADDING_CONFIGS = {
  none: 0,
  xs: 8,
  sm: 16,
  md: 24,
  lg: 32,
  xl: 40,
} as const;

export const CONTAINER_MARGIN_CONFIGS = {
  none: 0,
  xs: 8,
  sm: 16,
  md: 24,
  lg: 32,
  xl: 40,
  auto: 'auto',
} as const;

export const CONTAINER_ELEVATION_CONFIGS = {
  none: 0,
  subtle: 1,
  low: 2,
  medium: 4,
  high: 8,
  higher: 12,
  highest: 24,
} as const;

export const CONTAINER_BORDER_RADIUS = {
  none: 0,
  small: 4,
  medium: 8,
  large: 12,
  xlarge: 16,
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  defaultRole: 'main',
  sectionRole: 'section',
  regionRole: 'region',
  containerLabel: 'Content container',
} as const;

export const CONTAINER_VARIANT_CONFIGS = {
  fluid: {
    maxWidth: false,
    padding: CONTAINER_GUTTERS,
    centered: false,
    description: 'Full width container with gutters',
  },
  fixed: {
    maxWidth: 'lg' as ContainerMaxWidth,
    padding: CONTAINER_GUTTERS,
    centered: true,
    description: 'Fixed width container centered on page',
  },
  constrained: {
    maxWidth: 'md' as ContainerMaxWidth,
    padding: CONTAINER_GUTTERS,
    centered: true,
    description: 'Constrained width container for optimal reading',
  },
} as const;

export const RESPONSIVE_BEHAVIOR = {
  fluid: {
    xs: '100%',
    sm: '100%', 
    md: '100%',
    lg: '100%',
    xl: '100%',
  },
  fixed: {
    xs: '100%',
    sm: `${CONTAINER_MAX_WIDTH_VALUES.sm}px`,
    md: `${CONTAINER_MAX_WIDTH_VALUES.md}px`,
    lg: `${CONTAINER_MAX_WIDTH_VALUES.lg}px`,
    xl: `${CONTAINER_MAX_WIDTH_VALUES.xl}px`,
  },
  constrained: {
    xs: '100%',
    sm: `${CONTAINER_MAX_WIDTH_VALUES.sm}px`,
    md: `${CONTAINER_MAX_WIDTH_VALUES.md}px`,
    lg: `${CONTAINER_MAX_WIDTH_VALUES.md}px`,
    xl: `${CONTAINER_MAX_WIDTH_VALUES.md}px`,
  },
} as const;

export const DEFAULT_CONTAINER_PROPS = {
  variant: CONTAINER_VARIANTS.fixed,
  maxWidth: CONTAINER_MAX_WIDTHS.lg,
  centered: true,
  disableGutters: false,
  fluid: false,
  bordered: false,
  rounded: false,
  elevation: CONTAINER_ELEVATION_CONFIGS.none,
} as const;