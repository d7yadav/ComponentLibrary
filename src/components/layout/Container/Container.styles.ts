import { Container as MuiContainer } from '@mui/material';
import { styled } from '@mui/material/styles';
import React from 'react';

import {
  CONTAINER_MAX_WIDTH_VALUES,
  CONTAINER_GUTTERS,
  CONTAINER_BORDER_RADIUS,
  CONTAINER_ELEVATION_CONFIGS,
} from './Container.constants';
export const StyledContainer: React.ComponentType<any> = styled(MuiContainer, {
  shouldForwardProp: (prop) => !['customVariant', 'customMaxWidth', 'centered', 'bordered', 'rounded', 'customElevation', 'customBgcolor', 'customPadding', 'customMargin', 'customMinHeight', 'customMaxHeight'].includes(prop as string),
})<any>((props: any) => {
  const { 
    theme, 
    customVariant, 
    customMaxWidth, 
    centered, 
    bordered, 
    rounded, 
    customElevation, 
    customBgcolor,
    customPadding,
    customMargin,
    customMinHeight,
    customMaxHeight
  } = props;
  
  return {
    // Base container styles
    'width': '100%',
    'marginLeft': centered ? 'auto' : 0,
    'marginRight': centered ? 'auto' : 0,
    'display': 'block',
    'boxSizing': 'border-box',
    
    // Custom margin
    ...(customMargin && {
      'margin': typeof customMargin === 'number' ? theme.spacing(customMargin) : customMargin,
    }),
    
    // Custom padding
    ...(customPadding && {
      'padding': typeof customPadding === 'number' ? theme.spacing(customPadding) : customPadding,
    }),
    
    // Height constraints
    ...(customMinHeight && {
      'minHeight': typeof customMinHeight === 'number' ? `${customMinHeight}px` : customMinHeight,
    }),
    
    ...(customMaxHeight && {
      'maxHeight': typeof customMaxHeight === 'number' ? `${customMaxHeight}px` : customMaxHeight,
    }),
    
    // Background color
    ...(customBgcolor && {
      'backgroundColor': customBgcolor.includes('.') 
        ? theme.palette[customBgcolor.split('.')[0] as keyof typeof theme.palette]?.main || customBgcolor
        : customBgcolor,
    }),
    
    // Border
    ...(bordered && {
      'border': `1px solid ${theme.palette.divider}`,
    }),
    
    // Border radius
    ...(rounded && {
      'borderRadius': CONTAINER_BORDER_RADIUS.medium,
    }),
    
    // Elevation
    ...(customElevation && {
      'boxShadow': theme.shadows[Math.min(customElevation, 24)],
    }),
    
    // Variant-specific styles
    ...(customVariant === 'fluid' && {
      'maxWidth': 'none !important',
      [theme.breakpoints.up('xs')]: {
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.xs / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.xs / 8),
      },
      [theme.breakpoints.up('sm')]: {
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.sm / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.sm / 8),
      },
    }),
    
    ...(customMaxWidth && {
      [theme.breakpoints.up('xs')]: {
        'maxWidth': customMaxWidth === 'xs' ? CONTAINER_MAX_WIDTH_VALUES.xs : '100%',
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.xs / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.xs / 8),
      },
      [theme.breakpoints.up('sm')]: {
        'maxWidth': customMaxWidth === 'sm' || customMaxWidth === 'xs' ? CONTAINER_MAX_WIDTH_VALUES.sm : '100%',
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.sm / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.sm / 8),
      },
      [theme.breakpoints.up('md')]: {
        'maxWidth': ['md', 'sm', 'xs'].includes(customMaxWidth) ? CONTAINER_MAX_WIDTH_VALUES.md : '100%',
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.md / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.md / 8),
      },
      [theme.breakpoints.up('lg')]: {
        'maxWidth': ['lg', 'md', 'sm', 'xs'].includes(customMaxWidth) ? CONTAINER_MAX_WIDTH_VALUES.lg : '100%',
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.lg / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.lg / 8),
      },
      [theme.breakpoints.up('xl')]: {
        'maxWidth': customMaxWidth === 'xl' ? CONTAINER_MAX_WIDTH_VALUES.xl : 
                 ['lg', 'md', 'sm', 'xs'].includes(customMaxWidth) ? CONTAINER_MAX_WIDTH_VALUES[customMaxWidth] : '100%',
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.xl / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.xl / 8),
      },
    }),
    
    ...(customVariant === 'constrained' && {
      [theme.breakpoints.up('xs')]: {
        'maxWidth': CONTAINER_MAX_WIDTH_VALUES.xs,
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.xs / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.xs / 8),
      },
      [theme.breakpoints.up('sm')]: {
        'maxWidth': CONTAINER_MAX_WIDTH_VALUES.sm,
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.sm / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.sm / 8),
      },
      [theme.breakpoints.up('md')]: {
        'maxWidth': CONTAINER_MAX_WIDTH_VALUES.md,
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.md / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.md / 8),
      },
      [theme.breakpoints.up('lg')]: {
        'maxWidth': CONTAINER_MAX_WIDTH_VALUES.md, // Stay at md width
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.lg / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.lg / 8),
      },
      [theme.breakpoints.up('xl')]: {
        'maxWidth': CONTAINER_MAX_WIDTH_VALUES.md, // Stay at md width
        'paddingLeft': theme.spacing(CONTAINER_GUTTERS.xl / 8),
        'paddingRight': theme.spacing(CONTAINER_GUTTERS.xl / 8),
      },
    }),
  };
});

// Utility styled components for common container patterns
export const FluidContainer: React.ComponentType<any> = styled(StyledContainer)<any>(() => ({
  'maxWidth': 'none !important',
  'width': '100%',
}));

export const CenteredContainer: React.ComponentType<any> = styled(StyledContainer)<any>(() => ({
  'marginLeft': 'auto',
  'marginRight': 'auto',
}));

export const SectionContainer: React.ComponentType<any> = styled(StyledContainer, {
  shouldForwardProp: (prop) => !['spacing'].includes(prop as string),
})<{ spacing?: number }>((props: any) => {
  const { theme, spacing = 4 } = props;
  return {
    'paddingTop': theme.spacing(spacing),
    'paddingBottom': theme.spacing(spacing),
  };
});

export const HeroContainer: React.ComponentType<any> = styled(StyledContainer)<any>((props: any) => {
  const { theme } = props;
  return {
    'minHeight': '50vh',
    'display': 'flex',
    'alignItems': 'center',
    'justifyContent': 'center',
    'textAlign': 'center',
    'paddingTop': theme.spacing(8),
    'paddingBottom': theme.spacing(8),
  };
});

export const ContentContainer: React.ComponentType<any> = styled(StyledContainer)<any>((props: any) => {
  const { theme } = props;
  return {
    'paddingTop': theme.spacing(3),
    'paddingBottom': theme.spacing(3),
    [theme.breakpoints.up('md')]: {
      'paddingTop': theme.spacing(6),
      'paddingBottom': theme.spacing(6),
    },
  };
});