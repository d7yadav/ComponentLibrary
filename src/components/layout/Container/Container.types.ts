import type { ContainerProps as MuiContainerProps } from '@mui/material/Container';
import type { ReactNode } from 'react';

export type ContainerMaxWidth = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | false;

export type ContainerVariant = 'fluid' | 'fixed' | 'constrained';

export interface ContainerProps extends Omit<MuiContainerProps, 'maxWidth'> {
  children: ReactNode;
  
  variant?: ContainerVariant;
  
  maxWidth?: ContainerMaxWidth;
  
  centered?: boolean;
  
  disableGutters?: boolean;
  
  fluid?: boolean;
  
  padding?: number | string;
  
  margin?: number | string;
  
  bgcolor?: string;
  
  minHeight?: number | string;
  
  maxHeight?: number | string;
  
  bordered?: boolean;
  
  rounded?: boolean;
  
  elevation?: number;
  
  className?: string;
  
  sx?: object;
  
  'aria-label'?: string;
  
  'aria-describedby'?: string;
  
  role?: string;
}

export interface ContainerStyleProps {
  variant: ContainerVariant;
  maxWidth: ContainerMaxWidth;
  centered: boolean;
  disableGutters: boolean;
  fluid: boolean;
  bordered: boolean;
  rounded: boolean;
  elevation: number;
  bgcolor?: string;
  padding?: number | string;
  margin?: number | string;
  minHeight?: number | string;
  maxHeight?: number | string;
}

export interface ResponsiveBreakpoints {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
}

export interface ContainerConfiguration {
  maxWidths: Record<string, number>;
  
  gutters: Record<string, number>;
  
  breakpoints: ResponsiveBreakpoints;
}