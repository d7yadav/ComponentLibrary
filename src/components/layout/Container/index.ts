export { Container } from './Container';
export type { 
  ContainerProps, 
  ContainerMaxWidth, 
  ContainerVariant,
  ContainerStyleProps,
  ResponsiveBreakpoints,
  ContainerConfiguration
} from './Container.types';
export {
  CONTAINER_VARIANTS,
  CONTAINER_MAX_WIDTHS,
  CONTAINER_BREAKPOINTS,
  CONTAINER_MAX_WIDTH_VALUES,
  CONTAINER_GUTTERS,
  CONTAINER_PADDING_CONFIGS,
  CONTAINER_MARGIN_CONFIGS,
  CONTAINER_ELEVATION_CONFIGS,
  CONTAINER_BORDER_RADIUS,
  ACCESSIBILITY_CONSTANTS,
  CONTAINER_VARIANT_CONFIGS,
  RESPONSIVE_BEHAVIOR,
  DEFAULT_CONTAINER_PROPS,
} from './Container.constants';
export {
  FluidContainer,
  CenteredContainer,
  SectionContainer,
  HeroContainer,
  ContentContainer,
} from './Container.styles';