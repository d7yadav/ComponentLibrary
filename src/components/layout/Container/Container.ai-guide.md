# ü§ñ AI Coding Guide: Container

## Component Overview

**Category**: layout  
**Complexity**: medium  
**AI Suitability**: High - Well-structured for AI development

## Quick AI Reference

### Props Interface
No specific props defined.

### Available Variants
No specific variants defined.

## AI Implementation Guide

### Step-by-Step Implementation

1. **Import the component**:
   ```typescript
   import { Container } from '@/components/layout/Container';
   ```

2. **Define props interface** (if extending):
   ```typescript
   interface MyContainerProps extends ContainerProps {
     // Additional props
   }
   ```

3. **Implement with proper TypeScript**:
   ```typescript
   const MyContainer = (props: MyContainerProps) => {
     return <Container {...props} />;
   };
   ```

### AI-Specific Guidelines

- Always use TypeScript strict mode
- Implement proper prop validation
- Include accessibility attributes
- Follow component naming conventions
- Use absolute imports with @ prefix

## Common AI Coding Mistakes

### Mistakes to Avoid

1. **‚ùå Using default exports**
   ```typescript
   // Wrong
   export default Container;
   ```
   
   **‚úÖ Use named exports**
   ```typescript
   // Correct
   export { Container };
   ```

2. **‚ùå Missing TypeScript types**
   ```typescript
   // Wrong
   const props: any = { ... };
   ```
   
   **‚úÖ Proper typing**
   ```typescript
   // Correct
   const props: ContainerProps = { ... };
   ```

3. **‚ùå Inline styles without theme**
   ```typescript
   // Wrong
   <Container style={{ color: 'red' }} />
   ```
   
   **‚úÖ Theme-based styling**
   ```typescript
   // Correct
   <Container sx={{ color: 'error.main' }} />
   ```

4. **‚ùå Missing accessibility attributes**
   ```typescript
   // Wrong
   <Container onClick={handler} />
   ```
   
   **‚úÖ Proper accessibility**
   ```typescript
   // Correct
   <Container 
     onClick={handler}
     aria-label="Descriptive action"
     role="button"
   />
   ```

## AI Quality Checklist

### Pre-Implementation Checklist

- [ ] Component follows naming conventions (PascalCase)
- [ ] Props interface is properly typed with `ContainerProps`
- [ ] Component uses named exports only
- [ ] Absolute imports are used (`@/components/...`)
- [ ] Theme integration is implemented
- [ ] Accessibility attributes are included
- [ ] Component is responsive
- [ ] Error boundaries are considered
- [ ] Performance optimizations are applied

### Code Quality Checklist

- [ ] TypeScript strict mode compliance
- [ ] No `any` types used
- [ ] Proper error handling
- [ ] Consistent code formatting
- [ ] JSDoc documentation added
- [ ] Component is testable
- [ ] Storybook stories created
- [ ] Accessibility tested

### Performance Checklist

- [ ] React.memo used if appropriate
- [ ] Expensive operations memoized
- [ ] Bundle size impact considered
- [ ] Re-render optimization implemented
- [ ] Props destructuring optimized

## AI Testing Patterns

### Recommended Testing Approach

1. **Unit Tests**:
   ```typescript
   import { render, screen } from '@testing-library/react';
   import { Container } from './Container';
   
   describe('Container', () => {
     it('renders with default props', () => {
       render(<Container />);
       expect(screen.getByRole('...')).toBeInTheDocument();
     });
   });
   ```

2. **Props Testing**:
   ```typescript
   it('handles all prop combinations', () => {
     const props: ContainerProps = {
       // Test all required props
     };
     render(<Container {...props} />);
   });
   ```

3. **Accessibility Testing**:
   ```typescript
   import { axe, toHaveNoViolations } from 'jest-axe';
   
   it('meets accessibility standards', async () => {
     const { container } = render(<Container />);
     const results = await axe(container);
     expect(results).toHaveNoViolations();
   });
   ```

4. **Interaction Testing**:
   ```typescript
   import userEvent from '@testing-library/user-event';
   
   it('handles user interactions', async () => {
     const user = userEvent.setup();
     const handleClick = vi.fn();
     render(<Container onClick={handleClick} />);
     
     await user.click(screen.getByRole('...'));
     expect(handleClick).toHaveBeenCalled();
   });
   ```

## AI Recommendations

üß™ **Missing Tests**: Implement comprehensive test suite with >95% coverage
üìö **Missing Storybook**: Create comprehensive Storybook stories for documentation
‚ö° **Performance**: Use React.memo and useMemo for optimization
‚ôø **Accessibility**: Ensure WCAG 2.1 AA compliance
üé® **Theming**: Leverage theme system for consistent styling

---

*This guide is optimized for AI-assisted development. Follow these patterns for consistent, high-quality component implementation.*
