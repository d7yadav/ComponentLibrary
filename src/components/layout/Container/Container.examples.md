# ðŸ“– Container Examples

## Basic Usage

```typescript
import { Container } from '@/components/layout/Container';

// Basic usage
const Example = () => {
  return (
    <Container>
      Basic Container example
    </Container>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Container
      variant="primary"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Container
    </Container>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Container
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Container disabled>
    No content available
  </Container>
);

// Edge case: Error states
const ErrorState = () => (
  <Container 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Container 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedContainer = React.memo((props: ContainerProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Container 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Container key={index}>
          Item {index}
        </Container>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Container
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Container
    </Container>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Container
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Container>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Container component.*
