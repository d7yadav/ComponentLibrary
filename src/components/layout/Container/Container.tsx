import { forwardRef, memo } from 'react';

import {
  CONTAINER_VARIANTS,
  CONTAINER_MAX_WIDTHS,
  DEFAULT_CONTAINER_PROPS,
  ACCESSIBILITY_CONSTANTS,
} from './Container.constants';
import { StyledContainer } from './Container.styles';
import type { ContainerProps } from './Container.types';

/**
 * Enhanced Container component with responsive breakpoint support and flexible styling
 * 
 * Features:
 * - 3 variants: fluid, fixed, constrained
 * - Responsive breakpoint support (xs, sm, md, lg, xl)
 * - Flexible styling options: padding, margin, background, borders
 * - Elevation support with Material-UI shadows
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const Container = memo(forwardRef<HTMLDivElement, ContainerProps>((
  {
    children,
    variant = DEFAULT_CONTAINER_PROPS.variant,
    maxWidth = DEFAULT_CONTAINER_PROPS.maxWidth,
    centered = DEFAULT_CONTAINER_PROPS.centered,
    disableGutters = DEFAULT_CONTAINER_PROPS.disableGutters,
    fluid = DEFAULT_CONTAINER_PROPS.fluid,
    padding,
    margin,
    bgcolor,
    minHeight,
    maxHeight,
    bordered = DEFAULT_CONTAINER_PROPS.bordered,
    rounded = DEFAULT_CONTAINER_PROPS.rounded,
    elevation = DEFAULT_CONTAINER_PROPS.elevation,
    className,
    sx,
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role = ACCESSIBILITY_CONSTANTS.defaultRole,
    ...other
  },
  ref
) => {
  // Override maxWidth if fluid is true
  const effectiveMaxWidth = fluid ? false : maxWidth;
  
  // Override variant if fluid is true
  const effectiveVariant = fluid ? CONTAINER_VARIANTS.fluid : variant;
  
  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel || ACCESSIBILITY_CONSTANTS.containerLabel,
    'aria-describedby': ariaDescribedBy,
    role,
  };

  return (
    <StyledContainer data-testid="container"
      ref={ref}
      maxWidth={effectiveMaxWidth}
      disableGutters={disableGutters}
      className={className}
      sx={sx}
      // Custom props for styled component
      customVariant={effectiveVariant}
      customMaxWidth={effectiveMaxWidth}
      centered={centered}
      bordered={bordered}
      rounded={rounded}
      customElevation={elevation}
      customBgcolor={bgcolor}
      customPadding={padding}
      customMargin={margin}
      customMinHeight={minHeight}
      customMaxHeight={maxHeight}
      {...accessibilityProps}
      {...other}
    >
      {children}
    </StyledContainer>
  );
}));

Container.displayName = 'Container';