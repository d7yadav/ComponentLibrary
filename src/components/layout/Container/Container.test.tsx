/**
 * @fileoverview Container Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Container } from './Container';
import { CONTAINER_CONSTANTS } from './Container.constants';
import type { ContainerProps } from './Container.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Container Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Container>Test content</Container>);
      expect(screen.getByRole('main')).toBeInTheDocument();
          });

    it('renders children correctly', () => {
      const testContent = 'Test Container Content';
      renderWithTheme(<Container>{testContent}</Container>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-container';
      renderWithTheme(
        <Container data-testid={testId}>Test</Container>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Container className={customClass}>Test</Container>
      );
      expect(screen.getByRole('main')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: ContainerProps['variant'][] = [
        'fluid',
        'fixed',
        'constrained',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Container variant={variant} data-testid="variant-test">
              {variant} variant
            </Container>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: ContainerProps['size'][] = [
        'small',
        'medium',
        'large',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Container size={size} data-testid="size-test">
              {size} size
            </Container>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Container disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Container
        </Container>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events when clickable', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Container onClick={handleClick} data-testid="clickable-container">
          Clickable Container
        </Container>
      );
      
      const element = screen.getByTestId('clickable-container');
      fireEvent.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('does not interfere with child interactions', () => {
      const parentClick = vi.fn();
      const childClick = vi.fn();
      
      renderWithTheme(
        <Container onClick={parentClick} data-testid="parent-container">
          <button onClick={childClick} data-testid="child-button">
            Child Button
          </button>
        </Container>
      );
      
      const childElement = screen.getByTestId('child-button');
      fireEvent.click(childElement);
      expect(childClick).toHaveBeenCalledTimes(1);
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Container>Accessible Container</Container>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Container aria-label={ariaLabel}>
          Container for screen readers
        </Container>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Container>Keyboard Container</Container>);
      
      const element = screen.getByRole('main');
      element.focus();
      expect(element).toHaveFocus();
          });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Container 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Container
        </Container>
      );
      
      const element = screen.getByRole('main');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiContainer: {
            defaultProps: {
              variant: 'fluid',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Container data-testid="themed-container">
            Themed Container
          </Container>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-container');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Container>Dark Mode Container</Container>
        </ThemeProvider>
      );

      expect(screen.getByRole('main')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Container key={i}>Container {i}</Container>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="main"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Container />);
      expect(screen.getByRole('main')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Container>{null}</Container>);
      expect(screen.getByRole('main')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Container 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Container>
      );
      expect(screen.getByRole('main')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Container>Default Container</Container>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: ContainerProps['variant'][] = [
        'fluid',
        'fixed',
        'constrained',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Container variant={variant}>{variant} Container</Container>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
