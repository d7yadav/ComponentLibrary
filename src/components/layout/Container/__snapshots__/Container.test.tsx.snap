// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Container Component > Snapshots > matches snapshot for all variants > variant-constrained 1`] = `
<div
  aria-label="Content container"
  class="MuiContainer-root MuiContainer-maxWidthLg css-2g5gxw-MuiContainer-root"
  data-testid="container"
  role="main"
>
  constrained
   Container
</div>
`;

exports[`Container Component > Snapshots > matches snapshot for all variants > variant-fixed 1`] = `
<div
  aria-label="Content container"
  class="MuiContainer-root MuiContainer-maxWidthLg css-agtab2-MuiContainer-root"
  data-testid="container"
  role="main"
>
  fixed
   Container
</div>
`;

exports[`Container Component > Snapshots > matches snapshot for all variants > variant-fluid 1`] = `
<div
  aria-label="Content container"
  class="MuiContainer-root MuiContainer-maxWidthLg css-b3cund-MuiContainer-root"
  data-testid="container"
  role="main"
>
  fluid
   Container
</div>
`;

exports[`Container Component > Snapshots > matches snapshot for default props 1`] = `
<div
  aria-label="Content container"
  class="MuiContainer-root MuiContainer-maxWidthLg css-agtab2-MuiContainer-root"
  data-testid="container"
  role="main"
>
  Default Container
</div>
`;
